### PR TITLE
Patch: Document metaprogrammed module attributes

### PR DESCRIPTION
**Pull request by garrettheath4**
Expanded documentation to include dynamically-generated module-level attributes that cannot be detected (by an IDE) or documented (by NumpyDoc) without running the code. Specifically:

* `adafruit_logging.NOTSET`
* `adafruit_logging.DEBUG`
* `adafruit_logging.INFO`
* `adafruit_logging.WARNING`
* `adafruit_logging.ERROR`
* `adafruit_logging.CRITICAL`

are all module-level attributes that are generated using this block of metaprogramming (line 65):

```
LEVELS = [
    (00, "NOTSET"),
    (10, "DEBUG"),
    (20, "INFO"),
    (30, "WARNING"),
    (40, "ERROR"),
    (50, "CRITICAL"),
]

for level_value, level_name in LEVELS:
    globals()[level_name] = level_value
```

### CODE REVIEW

File: adafruit_logging.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 #
 # SPDX-License-Identifier: MIT
 
+# noinspection PyUnresolvedReferences
```
Comment (garrettheath4) [inline adafruit_logging.py:5]: `PyUnresolvedReferences` is a code inspection in Pycharm (by Jetbrains)

File: adafruit_logging.py [hunk 2]
```diff
@@ -41,54 +71,56 @@
     (50, "CRITICAL"),
 ]
 
-for value, name in LEVELS:
-    globals()[name] = value
+for __value, __name in LEVELS:
```
Comment (garrettheath4) [inline adafruit_logging.py:74]: Pycharm seems to think that these two lines of code define variables named "value" and "name" in the global scope. I'm not sure why. But adding the underscores in front makes sure that they're sorted at the bottom of code auto-completion results if so.

File: adafruit_logging.py (hunk 1 above)
Comment (brentru) [reply adafruit_logging.py:5]: Could you please remove this comment?

File: adafruit_logging.py [hunk 3]
```diff
@@ -30,6 +55,11 @@
 
 __version__ = "0.0.0-auto.0"
 __repo__ = "https://github.com/adafruit/Adafruit_CircuitPython_Logger.git"
+# noinspection PyUnresolvedReferences
```
Comment (brentru) [inline adafruit_logging.py:58]: Please remove this comment

File: adafruit_logging.py (hunk 2 above)
Comment (brentru) [reply adafruit_logging.py:74]: Could you keep the variable names the same?

File: adafruit_logging.py [hunk 4]
```diff
@@ -41,54 +71,56 @@
...
+def level_for(level_value):
+    """Convert a numeric level to the most appropriate name.
 
-    :param value: a numeric level
+    :param int level_value: a numeric level
 
     """
     for i in range(len(LEVELS)):
-        if value == LEVELS[i][0]:
+        if level_value == LEVELS[i][0]:
```
Comment (brentru) [inline adafruit_logging.py:85]: why did you change `value` to `level_value`? Is it conflicting with another variable name?

File: adafruit_logging.py [hunk 5]
```diff
@@ -143,91 +177,104 @@ def getEffectiveLevel(self):
...
 
-    def addHandler(self, hldr):
+    def addHandler(self, handler):
         """Sets the handler of this logger to the specified handler.
         *NOTE* this is slightly different from the CPython equivalent which adds
-        the handler rather than replaceing it.
+        the handler rather than replacing it.
 
-        :param hldr: the handler
+        :param handler: the handler
```
Comment (brentru) [inline adafruit_logging.py:185]: Since we're patching docstrings - would we also want to document the type of `handler` here?

File: adafruit_logging.py [hunk 6]
```diff
@@ -143,91 +177,104 @@ def getEffectiveLevel(self):
...
         """Log a debug message.
 
-        :param format_string: the core message string with embedded formatting directives
-        :param args: arguments to ``format_string.format()``, can be empty
+        :param str format_string: the core message string with embedded
+                                  formatting directives
+        :param args: arguments to ``format_string.format()``; can be empty
 
         """
+        # noinspection PyUnresolvedReferences
```
Comment (brentru) [inline adafruit_logging.py:210]: Please remove this comment

File: adafruit_logging.py [hunk 7]
```diff
@@ -143,91 +177,104 @@ def getEffectiveLevel(self):
...
         """Log a warning message.
 
-        :param format_string: the core message string with embedded formatting directives
-        :param args: arguments to ``format_string.format()``, can be empty
+        :param str format_string: the core message string with embedded
+                                  formatting directives
+        :param args: arguments to ``format_string.format()``; can be empty
 
         """
+        # noinspection PyUnresolvedReferences
```
Comment (brentru) [inline adafruit_logging.py:232]: Please remove this comment

File: adafruit_logging.py [hunk 8]
```diff
@@ -143,91 +177,104 @@ def getEffectiveLevel(self):
...
         """Log a info message.
 
-        :param format_string: the core message string with embedded formatting directives
-        :param args: arguments to ``format_string.format()``, can be empty
+        :param str format_string: the core message string with embedded
+                                  formatting directives
+        :param args: arguments to ``format_string.format()``; can be empty
 
         """
+        # noinspection PyUnresolvedReferences
```
Comment (brentru) [inline adafruit_logging.py:221]: Please remove this comment

File: adafruit_logging.py [hunk 9]
```diff
@@ -143,91 +177,104 @@ def getEffectiveLevel(self):
...
         """Log a error message.
 
-        :param format_string: the core message string with embedded formatting directives
-        :param args: arguments to ``format_string.format()``, can be empty
+        :param str format_string: the core message string with embedded
+                                  formatting directives
+        :param args: arguments to ``format_string.format()``; can be empty
 
         """
+        # noinspection PyUnresolvedReferences
```
Comment (brentru) [inline adafruit_logging.py:243]: Please remove this comment

File: adafruit_logging.py [hunk 10]
```diff
@@ -143,91 +177,104 @@ def getEffectiveLevel(self):
...
         """Log a critical message.
 
-        :param format_string: the core message string with embedded formatting directives
-        :param args: arguments to ``format_string.format()``, can be empty
+        :param str format_string: the core message string with embedded
+                                  formatting directives
+        :param args: arguments to ``format_string.format()``; can be empty
 
         """
+        # noinspection PyUnresolvedReferences
```
Comment (brentru) [inline adafruit_logging.py:254]: Please remove this comment

**Review by brentru — CHANGES_REQUESTED** [review]
Thank you for the PR- requesting a few changes.

File: adafruit_logging.py (hunk 5 above)
Comment (garrettheath4) [reply adafruit_logging.py:185]: Yes, I didn't understand what the type is supposed to be since the `Logger.addHandler(..)` method is not used anywhere in this file or the examples. But I now see it's supposed to be any subclass of the `LoggingHandler` abstract class.

**Review by garrettheath4 — COMMENTED** [review]
Replied to 2 comments and resolved suggestions by @brentru

Suggestions were to remove PyCharm-specific code inspection comments, add param type to `Logger.addHandler(..)` docstring, and revert `level_for(..)` param rename from `level_value` back to `value`

File: adafruit_logging.py (hunk 4 above)
Comment (brentru) [reply adafruit_logging.py]: Ok, thanks for the explanation, resolving

**Review by brentru — APPROVED** [review]

**Comment by brentru** [comment]
@garrettheath4  This is failing on `black`, could you fix this and re-commit. Changes look good.

**Comment by FoamyGuy** [comment]
I merged main to this branch and resolved the conflicts. Fixed one other pylint issue resulting from renaming some of the arguments in this PR.

Changes look good to me as well.
